pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -706,26 +710,64 @@ func (service *SecurityPolicyService) deleteSecurityPolicy(sp types.UID) error {
 		return err
 	}
 
-	err = securityPolicyStore.Apply(&finalSecurityPolicyCopy)
+	// Apply security policy and rules to store
+	err = service.applySecurityPolicyStore(finalSecurityPolicyCopy, finalSecurityPolicyCopy.Rules, *nsxGroupsEmpty, true)
 	if err != nil {
-		log.Error(err, "Failed to apply store", "securityPolicy", finalSecurityPolicyCopy)
 		return err
 	}
-	err = ruleStore.Apply(&finalSecurityPolicyCopy.Rules)
-	if err != nil {
-		log.Error(err, "Failed to apply store", "nsxRules", finalSecurityPolicyCopy.Rules)
-		return err
-	}
-	err = groupStore.Apply(nsxGroups)
+
+	// Delete groups after security policy and rules are deleted
+	err = service.deleteGroupsWithStore(nsxGroups, nil, groupStore)
 	if err != nil {
-		log.Error(err, "Failed to apply store", "nsxGroups", nsxGroups)
 		return err
 	}
 
 	log.Info("Successfully deleted NSX SecurityPolicy", "nsxSecurityPolicy", finalSecurityPolicyCopy)
 	return nil
 }
 
+// deleteGroupsWithStore deletes NSX groups from NSX and the group store.
+// It handles ResourceInUse errors by logging them but continuing execution.
+// Returns an error if deletion fails with a non-ResourceInUse error or if applying to the store fails.
+func (service *SecurityPolicyService) deleteGroupsWithStore(nsxGroups *[]model.Group, vpcInfo *common.VPCResourceInfo, groupStore *GroupStore) error {
+	if len(*nsxGroups) == 0 {
+		return nil
+	}
+
+	for _, group := range *nsxGroups {
+		groupID := *group.Id
+		var err error
+
+		// Use the appropriate client based on whether VPC is enabled
+		if IsVPCEnabled(service) {
+			err = service.NSXClient.VpcGroupClient.Delete(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, groupID)
+		} else {
+			failIfSubtreeExists, forceDelete := true, true
+			err = service.NSXClient.GroupClient.Delete(getDomain(service), groupID, &failIfSubtreeExists, &forceDelete)
+		}
+
+		err = nsxutil.TransNSXApiError(err)
+		if err != nil {
+			// Special handling for ResourceInUse errors - we can continue in this case
+			var nsxApiErr *nsxutil.NSXApiError
+			if errors.As(err, &nsxApiErr) && nsxApiErr.Type() == apierrors.ErrorType_RESOURCE_IN_USE {
+				log.Info("Some NSX groups are in use and cannot be deleted, continuing", "groupID", groupID)
```

Comment:
I think this log better to be a warning.
And if the group in use, the code doesn't do any, the group is still in the NSX and the NSX operator store; then if the group is not in use anymore, is the NSX operator able to collect it in GC and clean it?

Reply:
fixed